Mathlib/RingTheory/AlgebraicIndependent/Transcendental.lean

```diff
@@ -78,42 +83,53 @@ theorem AlgebraicIndependent.option_iff (hx : AlgebraicIndependent R x) (a : A)
+    (algebraicIndependent_empty_type_iff.mpr hinj) fun {a u} ha hu ha' h ↦ ?_
+  convert ((Set.image_eq_range _ _ ▸ h.option_iff <| x a).2 <| H u (hfin.subset hu) h _ ha').comp _
+    (Set.subtypeInsertEquivOption ha').injective with x
   by_cases h : ↑x = a <;> simp [h, Set.subtypeInsertEquivOption]
 
-/-- `Type` version of `algebraicIndependent_of_finite'`. -/
-theorem algebraicIndependent_of_finite_type'
+/-- Variant of `algebraicIndependent_of_finite` using `Transcendental`. -/
+theorem algebraicIndependent_of_finite' (s : Set A)
     (hinj : Injective (algebraMap R A))
-    (H : ∀ t : Set ι, t.Finite → ∀ i : ι, i ∉ t → Transcendental (adjoin R (x '' t)) (x i)) :
-    AlgebraicIndependent R x := by
-  nontriviality R
-  haveI := hinj.nontrivial
-  have hx : Injective x := by
-    simp_rw [Transcendental] at H
-    contrapose! H
-    obtain ⟨i, j, h1, h2⟩ := not_injective_iff.1 H
-    refine ⟨{j}, by simp, i, by simp [h2], ?_⟩
-    rw [h1, Set.image_singleton]
-    exact isAlgebraic_algebraMap (⟨x j, Algebra.self_mem_adjoin_singleton R _⟩ : adjoin R {x j})
-  rw [← Set.coe_comp_rangeFactorization x]
-  refine .comp (algebraicIndependent_of_finite' _ hinj fun t ht hfin a ha ha' ↦ ?_) _
-    (Set.rightInverse_rangeSplitting hx).injective
-  change Finite t at hfin
-  have := H (x ⁻¹' t) (Finite.of_injective _ (hx.restrictPreimage t))
-    ((Equiv.ofInjective _ hx).symm ⟨_, ha⟩)
-    (by rwa [Set.mem_preimage, Equiv.apply_ofInjective_symm hx])
-  rwa [Set.image_preimage_eq_inter_range, Set.inter_eq_self_of_subset_left ht,
-    Equiv.apply_ofInjective_symm hx] at this
+    (H : ∀ t ⊆ s, t.Finite → AlgebraicIndependent R ((↑) : t → A) →
+      ∀ a ∈ s, a ∉ t → Transcendental (adjoin R t) a) :
+    AlgebraicIndependent R ((↑) : s → A) :=
+  algebraicIndependent_of_finite_type' hinj fun t hfin h i hi ↦ H _
+    (by rintro _ ⟨x, _, rfl⟩; exact x.2) (hfin.image _) h.image _ i.2
+    (mt Subtype.val_injective.mem_set_image.mp hi)
+
+namespace AlgebraicIndependent
+
+variable (hx : AlgebraicIndependent R x)
+include hx
+
+theorem adjoin_of_disjoint {s t : Set ι} (h : Disjoint s t) :
+    AlgebraicIndependent (adjoin R (x '' s)) fun i : t ↦ x i := by
+  let e := (sumAlgEquiv R t s).trans (mapAlgEquiv t (hx.comp _ Subtype.val_injective).aevalEquiv)
+  have : ((aeval fun i : t ↦ x i).restrictScalars R).comp e.toAlgHom =
+      (aeval x).comp (rename <| Sum.elim Subtype.val Subtype.val) := by
+    ext (_|_) <;> simp [e, algebraMap_aevalEquiv]
+  have _ := @MvPolynomial.isScalarTower
+  rw [Set.image_eq_range, AlgebraicIndependent, ← AlgHom.coe_restrictScalars' R, ← e.injective_comp]
+  show Injective ((AlgHom.restrictScalars R <| aeval _).comp e.toAlgHom)
+  rw [this, AlgHom.coe_comp]
+  exact .comp hx (rename_injective _ <| Subtype.val_injective.sum_elim
+    Subtype.val_injective fun i j eq ↦ h.ne_of_mem j.2 i.2 eq.symm)
+
+theorem transcendental_adjoin {s : Set ι} {i : ι} (hi : i ∉ s) :
+    Transcendental (adjoin R (x '' s)) (x i) := by
```

Comment:
This and the above `adjoin_of_disjoint` should also have a `IntermediateField.adjoin` version. Would you like to add them?

Also, `transcendental_adjoin` should also have an `iff` version, which could be a `simp` lemma.

Maybe `adjoin_of_disjoint` also have an `iff` version?

I'm thinking of applications to `MvPolynomial` to obtain similar results in https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic/MvPolynomial.html but for `AlgebraicIndependent`.

Reply:
> Also, `transcendental_adjoin` should also have an `iff` version, which could be a `simp` lemma. Maybe `adjoin_of_disjoint` also have an `iff` version?

I added these, but `simp` lemmas with an `AlgebraicIndependent` hypothesis seem weird, so I didn't mark them `@[simp]`.

> This and the above `adjoin_of_disjoint` should also have a `IntermediateField.adjoin` version. Would you like to add them?

They don't fit in this file due to required import, so not in this PR.

Reply:
> > Also, `transcendental_adjoin` should also have an `iff` version, which could be a `simp` lemma. Maybe `adjoin_of_disjoint` also have an `iff` version?
> 
> I added these, but `simp` lemmas with an `AlgebraicIndependent` hypothesis seem weird, so I didn't mark them `@[simp]`.

Oh I forgot it. That's OK.
 
> > This and the above `adjoin_of_disjoint` should also have a `IntermediateField.adjoin` version. Would you like to add them?
> 
> They don't fit in this file due to required import, so not in this PR.

I see. Maybe later I'll add them to another file if they are needed.